src/libraries/System.Linq/src/System/Linq/Grouping.cs

```diff
@@ -118,7 +118,229 @@ public static partial class Enumerable
                 return [];
             }
 
-            return new GroupedResultEnumerable<TSource, TKey, TElement, TResult>(source, keySelector, elementSelector, resultSelector, comparer);
+            return new GroupByResultIterator<TSource, TKey, TElement, TResult>(source, keySelector, elementSelector, resultSelector, comparer);
+        }
+
+        internal sealed partial class GroupByResultIterator<TSource, TKey, TElement, TResult> : Iterator<TResult>
+        {
+            private readonly IEnumerable<TSource> _source;
+            private readonly Func<TSource, TKey> _keySelector;
+            private readonly Func<TSource, TElement> _elementSelector;
+            private readonly IEqualityComparer<TKey>? _comparer;
+            private readonly Func<TKey, IEnumerable<TElement>, TResult> _resultSelector;
+
+            private Lookup<TKey, TElement>? _lookup;
+            private Grouping<TKey, TElement>? _g;
+
+            public GroupByResultIterator(IEnumerable<TSource> source, Func<TSource, TKey> keySelector, Func<TSource, TElement> elementSelector, Func<TKey, IEnumerable<TElement>, TResult> resultSelector, IEqualityComparer<TKey>? comparer)
+            {
+                _source = source;
+                _keySelector = keySelector;
+                _elementSelector = elementSelector;
+                _comparer = comparer;
+                _resultSelector = resultSelector;
+            }
+
+            public override Iterator<TResult> Clone() => new GroupByResultIterator<TSource, TKey, TElement, TResult>(_source, _keySelector, _elementSelector, _resultSelector, _comparer);
+
+            public override bool MoveNext()
```

Comment:
Any perf improvements in `GroupBy` because of this change?

Reply:
The main reason for the change is these types were previously implementing the IIListProvider interface, and now that those methods have moved down to Iterator, to keep those optimizations these types needed to derive from Iterator.  But, yes, beyond that, it picks up some minor benefits, including avoiding allocating an extra enumerator object when you iterate through the GroupBy enumerable.

```C#
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;

BenchmarkSwitcher.FromAssembly(typeof(Tests).Assembly).Run(args);

[MemoryDiagnoser(false)]
[HideColumns("Job", "Error", "StdDev", "Median", "RatioSD")]
public partial class Tests
{
    private IEnumerable<int> _data = Enumerable.Range(0, 10);

    [Benchmark]
    public int GroupBy()
    {
        int sum = 0;
        foreach (var g in _data.GroupBy(x => x)) sum += g.Key;
        return sum;
    }
}
```

| Method  | Toolchain         | Mean     | Ratio | Allocated | Alloc Ratio |
|-------- |------------------ |---------:|------:|----------:|------------:|
| GroupBy | \main\corerun.exe | 470.8 ns |  1.00 |   1.25 KB |        1.00 |
| GroupBy | \pr\corerun.exe   | 428.6 ns |  0.91 |   1.23 KB |        0.99 |